### PR TITLE
Sync head

### PR DESCRIPTION
**Pull request by shayne-fletcher**
- Sync to https://gitlab.haskell.org/ghc/ghc.git `057db94ce038970b14df1599fe83097c284b9c1f`
- Update the `stack.yaml` resolver to be consistent with the `hadrian/stack.yaml` resolver;
- Factor out `applyPatchHadrianStackYaml` and add a line for happy (https://gitlab.haskell.org/ghc/ghc/-/issues/18726);
- ~Needs rebasing on master after https://github.com/digital-asset/ghc-lib/commit/8c2dae9a2026b32b1d4782affd700e493fa3cff9 lands.~

### CODE REVIEW

File: CI.hs [hunk 1]
```diff
@@ -176,6 +176,33 @@ parseStackOptions = StackOptions
...
+-- https://gitlab.haskell.org/ghc/ghc/-/commit/30272412fa437ab8e7a8035db94a278e10513413
+-- (4th May 2020), certain GHC modules depend on the exceptions
+-- package. Depending on the boot compiler, this package may or may
+-- not be present and if it's missing, determining the list of
+-- ghc-lib-parser modules (via 'calcParserModules') will fail. The
+-- function 'appyPatchHadrianStackYaml' guarantees it is available
+-- if its needed.
+
+-- Patch Hadrian's Cabal.
+-- TODO (SF, 2020-09-20): Move this function and the above note to
```
Comment (cocreature) [inline CI.hs:191]: Any reason not do this now? This seems pretty crucial. We don’t want to force people to use `CI`. `ghc-lib-gen` should work on its own.

Comment (shayne-fletcher) [reply CI.hs:191]: there's no deep reason just that i spotted in the course of this work it's in the wrong place. sure, let's knock it over. stand by.

**Comment by shayne-fletcher** [comment]
When https://gitlab.haskell.org/ghc/ghc/-/merge_requests/4096 containing https://gitlab.haskell.org/ghc/ghc/-/merge_requests/4096/diffs?commit_id=e436b73f3687b7cbca3f2ab63d9d753660aabbce lands, we'll be able to unwind writing the extra dep.

File: CI.hs (hunk 1 above)
Comment (shayne-fletcher) [reply CI.hs]: done

File: CI.hs [hunk 2]
```diff
@@ -258,8 +244,6 @@ buildDists
 
     -- Make and extract an sdist of ghc-lib.
     cmd "cd ghc && git checkout ."
-    appendFile "ghc/hadrian/stack.yaml" $ unlines ["ghc-options:","  \"$everything\": -O0 -j"]
-    -- Feedback on the compiler used for ghc-lib-gen.
```
Comment (hurryabit) [inline CI.hs:262]: We didn't add any extra-deps here but we seem to add them globally with `applyPatchHadrianStackYaml` now. Is this intended?

Comment (shayne-fletcher) [reply CI.hs:262]: The extra-deps were added a little bit further down: https://github.com/digital-asset/ghc-lib/blob/6be486b61f07c022402d632e161ab081c837a74b/CI.hs#L247-L249. This block is now correctly moved to ghc-lib-gen and an additional extra dep added (which is temporary while the upstream fix isn't landed).

Comment (hurryabit) [reply CI.hs:262]: The line you're referring to is a bit further up. Let me rephrase what I find confusing: Before this change we had tw locations where we wrote `ghc-options` to `stack.yaml` and only location where we wrote `extra-deps`. Now, we have one location where we write both. What happened to the case where we need to write the `ghc-options` but not the `extra-deps`?

Comment (shayne-fletcher) [reply CI.hs:262]: Before in CI.hs, we unconditionally wrote ghc-options and wrote extra-deps conditionally on flavor ghc-master in two consecutive statements. These two statements have been moved to ghc-lib-gen and the interior of the function `applyPatchHadrianStackYaml`. The function will not write extra-deps if the flavor is not ghc-master just as it were before.

At the risk of confusing things further but in the interests of completeness, there is one further change that is being overlooked: CI.hs used to patch hadrian.cabal exactly once. This was a (then not,  but now important) error. It should happen twice because in-between generating ghc-lib-parser.cabal and ghc-lib.cabal, the changes to hadrian/hadrian.cabal are reverted. Now, `Main.hs` arranges to patch  hadrian's cabal  on every execution of ghc-lib-gen regardless of whether the target is ghc-lib-parser or ghc-lib.

Comment (shayne-fletcher) [reply CI.hs:262]: In any case, the key point is that the two statements that were in CI.hs are exactly next two each other in ghc-lib-gen now as they were before https://github.com/shayne-fletcher/ghc-lib/blob/03e3d68df6fcd132fe7a9b836c9cd90b0616cde0/ghc-lib-gen/src/Ghclibgen.hs#L584-L594.

Comment (shayne-fletcher) [reply CI.hs:262]: @hurryabit  Ah! I see what you are saying. Yes, we were only writing the extra-deps once before. That is what I am wittering on about here https://github.com/digital-asset/ghc-lib/pull/245#discussion_r492125958.  We are agreed that this was not correct and this PR incidentally fixes that.

Comment (hurryabit) [reply CI.hs:262]: That explains it. Thanks a lot.

**Review by hurryabit — APPROVED** [review]
Looks good to me. Thank you.

**Comment by shayne-fletcher** [comment]
I've never seen anything like that Windows failure before. I suspect it will work if restarted.

**Comment by cocreature** [comment]
We got the same windows error on the daml CI today and solved it by upgrading stack.

**Comment by shayne-fletcher** [comment]
We can put a line in `azure-pipelines.yml`:
```
  - script: |
      curl -sSL https://get.haskellstack.org/ | sh
      stack upgrade
```
The effect then should be to guarantee stack is installed and that it is the most recent stack available. I'll push that and let's see.

**Comment by shayne-fletcher** [comment]
I think we are actually looking at the systems of a stack outage. On my clone all builds are failing right now with the same issue. Re-running at this time probably won't help. I think we have to wait.
